worker/dbaccessor/tracker.go

```diff
@@ -0,0 +1,244 @@
+	return func(w *trackedDBWorker) {
+		w.verifyDBFunc = f
+	}
+}
+
+// WithClock sets the clock used by the worker.
+func WithClock(clock clock.Clock) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.clock = clock
+	}
+}
+
+// WithLogger sets the logger used by the worker.
+func WithLogger(logger Logger) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.logger = logger
+	}
+}
+
+type trackedDBWorker struct {
+	tomb tomb.Tomb
+
+	dbApp     DBApp
+	namespace string
+
+	mutex sync.RWMutex
+	db    *sql.DB
+	err   error
+
+	clock  clock.Clock
+	logger Logger
+
+	verifyDBFunc func(context.Context, *sql.DB) error
+}
+
+// NewTrackedDBWorker creates a new TrackedDBWorker
+func NewTrackedDBWorker(dbApp DBApp, namespace string, opts ...TrackedDBWorkerOption) (TrackedDB, error) {
+	w := &trackedDBWorker{
+		dbApp:        dbApp,
+		namespace:    namespace,
+		clock:        clock.WallClock,
+		verifyDBFunc: defaultVerifyDBFunc,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	var err error
+	w.db, err = w.dbApp.Open(context.TODO(), w.namespace)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w.tomb.Go(w.loop)
+
+	return w, nil
+}
+
+// DB closes over a raw *sql.DB. Closing over the DB allows the late
+// realization of the database. Allowing retries of DB acquisition if there
+// is a failure that is non-retryable.
+func (w *trackedDBWorker) DB(fn func(*sql.DB) error) error {
+	w.mutex.RLock()
+	// We have a fatal error, the DB can not be accessed.
+	if w.err != nil {
+		w.mutex.RUnlock()
+		return errors.Trace(w.err)
+	}
+	db := w.db
+	w.mutex.RUnlock()
+
+	return fn(db)
+}
+
+// Txn closes over a raw *sql.Tx. This allows retry semantics in only one
+// location. For instances where the underlying sql database is busy or if
+// it's a common retryable error that can be handled cleanly in one place.
+func (w *trackedDBWorker) Txn(ctx context.Context, fn func(context.Context, *sql.Tx) error) error {
+	return w.DB(func(db *sql.DB) error {
+		return database.Retry(ctx, func() error {
+			return database.Txn(ctx, db, fn)
+		})
+	})
+}
+
+// Err will return any fatal errors that have occurred on the worker, trying
+// to acquire the database.
+func (w *trackedDBWorker) Err() error {
+	w.mutex.RLock()
+	defer w.mutex.RUnlock()
+
+	return w.err
+}
+
+// Kill implements worker.Worker
+func (w *trackedDBWorker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker
+func (w *trackedDBWorker) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *trackedDBWorker) loop() error {
+	timer := w.clock.NewTimer(PollInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-timer.Chan():
+			// Any retryable errors are handled at the txn level. If we get an
+			// error returning here, we've either exhausted the number of
+			// retries or the error was fatal.
+			w.mutex.RLock()
+			currentDB := w.db
+			w.mutex.RUnlock()
+
+			newDB, err := w.verifyDB(currentDB)
```

Comment:
hm. 'verifyDB' made me think it was Validation, and this had me concerned that we were validating the content every 10s. It looks like the default is just PingContext, which seems light weight enough.